### ISSUE TITLE
Add a rule to check uninitialized class properties under strictNullChecks

### ISSUE THREAD

**Issue by nippur72** [palantir/tslint#1414]
Under `--strictNullChecks`, class properties that are left uninitialized are problematic: 

``` ts
class Foo {
   someprop: string;
}
var a = new Foo();
a.someprop === undefined;  // true
```

I suggest we have a rule that checks that properties are initialized, either inline or in the constructor.

As of today, the equivalent compiler feature is marked **wontfix**, see https://github.com/Microsoft/TypeScript/issues/8476.

**Comment by adidahiya** [comment]
@nippur72 cool, that sounds like a valid rule to exist in core. Maybe something like "no-uninitialized-member".

**Comment by RyanCavanaugh** [comment]
How should this rule work for patterns like this?
```ts
class Foo {
  someprop: string;

  constructor() {
    this.init('self');
  }

  init(s: string) {
    this.someprop = s;
  }
}
```

**Comment by nippur72** [comment]
Ideally it should walk all paths in the constructor and check for `this` at the end of the constructor. 

But still there will be cases where the rule is going to fail, e.g.:
```ts
class Foo {
  someprop: string;

  constructor() {
    let a = this;
    a.someprop = "42";
  }
}
```

**Comment by helmutschneider** [comment]
[Try this](https://gist.github.com/helmutschneider/6ed2ed6aedd91ce8654efa1a710c7c25). At the moment it does not care about delegating initializers and/or base classes.

**Comment by abierbaum** [comment]
@helmutschneider Have you extended this rule at all since October?  I would love to see something like this in tslint to make strictNullChecks useful.

**Comment by helmutschneider** [comment]
@abierbaum I have not, no. It needs to be updated for tslint 4.0 but I believe that's a small change. Even if it were to be merged I'm not sure I would have the time to maintain it in the future. Feel free to use the code as you like - I give you my full consent.

**Comment by helmutschneider** [comment]
@abierbaum I just updated [the gist](https://gist.github.com/helmutschneider/6ed2ed6aedd91ce8654efa1a710c7c25) for tslint 4.0.

**Comment by abierbaum** [comment]
See https://github.com/alhugone/tslint-strict-null-checks for a potential way to implement this.

**Comment by dweber019** [comment]
Typescript has implemented this in version 2.7. https://blogs.msdn.microsoft.com/typescript/2018/01/31/announcing-typescript-2-7/

**Comment by adidahiya** [comment]
Closing in favor of the compiler feature
